### PR TITLE
fix(providers): guard native agent loops against mid-turn overflow

### PR DESCRIPTION
**Pull request by pdogra1299**
## Description

Two native provider loops grow their own history as the model calls tools, and neither had any in-turn context guard. A long agentic run overflowed the model's window mid-loop and lost every completed step.

Verified coverage before this PR (read from each loop, not grepped):

| Loop | In-turn guard |
|---|---|
| AI-SDK `generateText` | compact-and-continue |
| `googleVertex` | stop-only |
| `googleNativeGemini3` | stop-only |
| **`openaiChatCompletionsBase`** | **none** — LiteLLM, DeepSeek, OpenRouter, NVIDIA NIM, LM Studio, llama.cpp, openai-compatible |
| **`anthropic` (native stream loop)** | **none** |
| `googleAiStudio` | none — *not addressed here, see Scope* |

## Type of Change

- [x] Bug fix (non-breaking change which fixes an issue)

## Changes

- **`context/loopGuardCore.ts`** — provider-neutral reclaim policy. Each loop keeps its history in a different shape (AI-SDK `ModelMessage`, OpenAI-compatible `{role,tool_calls}`, Anthropic content blocks) but the policy is identical, so adapters map onto a neutral `LoopGuardEntry[]`, ask the core what to do, and apply the answer in their own types.
- **`context/openaiCompatLoopGuard.ts`** + wiring — `tool_call_id` pairing preserved, head/tail previews at the same 2KB/60-line budget the AI-SDK guard uses.
- **`context/anthropicLoopGuard.ts`** + wiring — granularity differs: one user message carries *every* `tool_result` for a step, so the message is the batch unit. Returns a PLAN and the loop applies it in its own `MessageParam` types, which avoids generic-spread assignability problems and the double assertions Critical Rule 14 forbids.
- Both wired to the provider's real reported prompt tokens for free calibration.

Invariants the policy never violates: entry 0 (the task) is never touched, and a tool batch is dropped whole or not at all — an orphaned `tool_call` is a hard provider rejection.

## Testing

Unit, no API keys:
- `pnpm run test:loop-guard-core` — **11/11**
- `pnpm run test:openai-compat-guard` — **9/9**
- `pnpm run test:anthropic-guard` — **10/10**

All three verified to fail **loudly** (`✗`, exit 1) on a deliberately broken assertion, per the CLAUDE.md hazard where `isExpectedProviderError()` can downgrade a real failure to `⊘`.

**Live, against a real Claude endpoint** (local Anthropic proxy):
- baseline agentic loop: 4 tool calls, correct answer, `cacheReadTokens: 53,247` at **100% cache savings** — proving the guard returned `undefined` and left the prefix byte-identical when the loop fit
- forced-firing run through `stream()`: guard fired 4×, `messagesDropped: 2` each time (one complete call+result batch), `calibration: 1.10` derived from the provider's real prompt tokens — the expected ~10% gap between the char estimate and the real tokenizer
- `toolResultsTruncated: 0` alongside drops confirms the core skips truncating entries it is about to drop

Regressions: `test:tool-pairing` 15/15, `test:token-accounting` 10/10, `test:step-guard` 6/6. `tsc --noEmit --strict` 0 errors, `eslint` 0 errors, prettier clean.

## Notes for reviewers

**`generate()` and `stream()` take different paths on Anthropic.** `generate()` routes through the AI-SDK `GenerationHandler`; the native loop in `providers/anthropic/client.ts` serves `stream()`. Both are now guarded — the former by the existing step guard, the latter by this PR — but that split is worth knowing when reading the code.

**Scope:** `googleAiStudio` is deliberately not wired, and `googleVertex` / `googleNativeGemini3` are not upgraded from stop-only. There are no AI-Studio credentials in this environment to verify against, and stop-only is degraded rather than broken — shipping those verified-by-construction only would be a worse trade than holding them.

Top of the stack: #1280 → #1281 → #1282 → this.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added automatic context protection for Anthropic and OpenAI-compatible tool loops.
  * Oversized tool results can be shortened, while older tool-call/result exchanges may be removed when necessary.
  * Preserves the original task and valid message pairing, with clear elision notes when history is reclaimed.
  * Improves token estimates using observed provider usage.

* **Tests**
  * Added continuous, network-free coverage for context limits, truncation, history removal, pairing, and calibration behavior.

### CODE REVIEW

**Comment by murdore** [comment]
Not mergeable yet, and I want to flag *why* the checks look the way they do, because the picture is misleading.

## Base is not `release`

This targets `fix/cache-aware-compaction` (#1282), so it can only merge after everything below it. Order for the stack:

```
release
 └─ #1280  fix/tool-pair-identity
     └─ #1281  fix/token-accounting
         └─ #1282  fix/cache-aware-compaction
             └─ #1283  fix/loop-guard-parity
                 └─ #1285  fix/session-persistence
                     └─ #1286  fix/redis-append-only
                         └─ #1287  fix/loop-guard-completion
```

This PR is **#4 of 7**.

## The Single Commit failure is a stacking artifact, not a real violation

`single-commit-enforcement.yml` triggers on `push` with `branches-ignore: [release, main, develop]`, so it runs on the branch and measures against `release` rather than against this PR's own base. `fix/cache-aware-compaction`… this branch is 4 commits ahead of `release` because it carries its parents'. Against its **own** base it is a single commit, which is what the policy actually asks for. It should clear itself as the parents merge.

## The bigger issue: almost no CI has run here

Three checks are present on this PR. Release-targeting PRs get **fifteen**. The difference is not incidental — the workflows are branch-gated:

| workflow | `pull_request: branches:` |
|---|---|
| `ci.yml` (`test (20)`, `build-check`, lint, format) | `[release]` |
| `yama-review.yml` | `[release]` |
| `copilot-review.yml` | `[release, main]` |

So on this PR **nothing has been built, no tests have run, and neither AI reviewer has looked at it.** Only CodeRabbit, SentinelOne and the push-triggered commit check fire. The absence of review comments here reflects the absence of a review, not a clean bill of health.

Across #1281, #1282, #1283, #1285, #1286 and #1287 that is roughly **3,900 added lines that have never been compiled or tested by CI**.

Nothing to do on this PR specifically — it resolves itself once the parents land and the base becomes `release`. Worth knowing before anyone reads the green ticks as coverage. If you want real signal sooner, temporarily retargeting to `release` (or adding the stack branches to `ci.yml`'s filter) would run the full gate.

**Comment by pdogra1299** [comment]
@murdore — confirmed both points, verified rather than taken on trust.

**Single Commit failure is a stacking artifact.** `single-commit-enforcement.yml` fires on `push` with `branches-ignore: [release, main, develop]`, and in the push branch of that script `BASE_BRANCH` is hard-coded to `origin/release` — so it measures against `release`, not the PR's own base. Against its own base every branch in the stack is exactly one commit:

```
$ git rev-list --count HEAD ^<own-base>
fix/token-accounting        1
fix/cache-aware-compaction  1
fix/loop-guard-parity       1
fix/session-persistence     1
fix/redis-append-only       1
fix/loop-guard-completion   1
```

Nothing to squash. It clears itself as parents land.

**The CI gap is real and I am not treating the green ticks as coverage.** Verified the gating you listed:

| workflow | `pull_request: branches:` |
|---|---|
| `ci.yml` | `[release]` |
| `yama-review.yml` | `[release]` |
| `copilot-review.yml` | `[release, main]` |

So on this PR nothing has been built or tested by CI, and that is the accurate reading.

What has run, locally on the **stack tip** (so every branch's code compiled together) after the latest rebase:

- `pnpm run check` → 4961 files, **0 errors, 0 warnings**
- all nine new suites → **95 tests, 0 failures, 0 skips**
  (`tool-pairing` 18, `token-accounting` 10, `step-guard` 6, `loop-guard-core` 11, `openai-compat-guard` 9, `anthropic-guard` 10, `tool-storage-parity` 10, `redis-append-only` 10, `gemini-guard` 11)

That is local signal, not CI signal, and I would rather say so plainly than let the tick count imply otherwise. Merging bottom-up closes the gap for real: each PR's base becomes `release` as its parent lands, at which point it picks up the full fifteen-check gate before it can merge — so no branch reaches `release` without having been through it.

**Comment by pdogra1299** [comment]
#1280 is merged (`a0f49e40`), so the stack has been rebased onto the new `release` (now at 10.10.2 — it had moved two releases since these branches were cut). Still one commit per branch.

**#1281 was `CONFLICTING` after the merge** — its branch still carried the pre-squash copy of #1280's commit. Replanted with `git rebase --onto origin/release 8344a885` so only its own commit survives; now `MERGEABLE` against `release`.

**One real conflict, in `providers/anthropic/client.ts`.** `release` had gained the structured-output buffering from the `final_result` work (`bufferedText` / `finalResultText`) at the same point where #1283 adds `estimateAnthropicFixedOverhead`. Both-added, both wanted — kept both. Verified afterwards that the release-side path is still fully wired (`bufferedText` accumulates at the text-delta branch, `finalResultText` is set from the `final_result` tool call, and the buffered turn is still delivered at the end of the loop), not just that it compiles.

## The build was broken, and `pnpm run check` did not catch it

Worth recording since it is exactly the CI gap @murdore flagged. Building the tip surfaced two errors in #1286's code that the typechecker passes clean:

```
redisConversationMemoryManager.ts(987,7): error TS2345:
  '(string | Buffer)[]' is not assignable to 'string[]'
redisConversationMemoryManager.ts(1000,5): error TS2322:
  'number | `${number}`' is not assignable to 'number'
```

`pnpm run check` uses the main tsconfig; `build:react-hooks` compiles with `--module nodenext --moduleResolution nodenext`, which resolves the `redis` types differently and widens both return types. Fixed at the source rather than with an assertion (Rule 14): `lRange` entries are normalized via `.toString()`, `lLen` through `Number()`. Both are real defects, not just type noise — a binary-mode client would have thrown on parse, and a RESP3 string count would have concatenated where callers add.

## Verification

Each branch was built **independently**, since each merges on its own:

| branch | build | 
|---|---|
| `fix/token-accounting` | ✅ |
| `fix/cache-aware-compaction` | ✅ |
| `fix/loop-guard-parity` | ✅ |
| `fix/session-persistence` | ✅ |
| `fix/redis-append-only` | ✅ |
| `fix/loop-guard-completion` | ✅ |

On the tip: `pnpm run check` → 4939 files, **0 errors**; `pnpm run lint` → **0 errors**, 48 warnings, all pre-existing `max-lines-per-function` in files this stack does not touch (`claudeProxyRoutes.ts`, `toolRoutes.ts`, `csvProcessor.ts`); all nine suites → **95 tests, 0 failures, 0 skips**.

Still local signal, not CI signal — except on #1281, which now targets `release` and is running the full gate.

**Comment by pdogra1299** [comment]
#1281 is merged (`613709b3`), so `release` is now at both stack heads #1280 and #1281. The remaining stack has been rebased onto it.

**Rebase — one commit each, no conflicts:**

```
94370cfb  #1287  fix(providers): complete in-turn context guard coverage across native loops
31f043cd  #1286  fix(memory): store session messages in an append-only redis list
ca47f289  #1285  fix(memory): persist tool executions on every memory backend
451fe2ba  #1283  fix(providers): guard native agent loops against mid-turn overflow
f722c458  #1282  fix(context): reclaim to a low-water mark in the step budget guard
613709b3  release
```

#1282 was `DIRTY` because it still carried the pre-squash copy of #1281's commit; that is resolved.

**Verification — each branch independently, not just the tip:**

| branch | `pnpm run check` | `pnpm run build` | own suites |
| --- | --- | --- | --- |
| `fix/cache-aware-compaction` | 4935 files, 0 errors | pass | step-guard 6/6 |
| `fix/loop-guard-parity` | 4938 files, 0 errors | pass | anthropic-guard 10/10, loop-guard-core 11/11, openai-compat-guard 9/9 |
| `fix/session-persistence` | 4938 files, 0 errors | pass | tool-storage-parity 10/10 |
| `fix/redis-append-only` | 4938 files, 0 errors | pass | redis-append-only 10/10 |
| `fix/loop-guard-completion` | 4939 files, 0 errors | pass | gemini-guard 11/11 |

67 tests, 0 failures, 0 skips. `pnpm run build` is run in full because `pnpm run check` alone does not catch everything — the `nodenext` resolution used by `build:react-hooks` surfaced two real `redis` typing defects in #1286 that `check` passed over.

**Review comments:** no unresolved inline findings on this PR. The only review activity to date is CodeRabbit's auto-generated walkthrough and @murdore's note about the check count, which is answered above — that gap closes as each PR's base becomes `release`. #1281 bore this out: once it targeted `release` it ran all 16 checks and passed, and that run triggered a CodeRabbit pass with three findings. Those landed after #1281 was already merged, so they are carried in a separate follow-up rather than retro-fitted here.

**Comment by Tara-ag** [comment]
🔍 MINOR: Calibration threshold parameter needs documentation - The observedPromptCount threshold parameter (default 200_000) lacks inline documentation explaining why this value was chosen and how it affects different model contexts. Add JSDoc comment explaining the calibration threshold: "When observed prompt count exceeds X tokens, use historical usage to calibrate the estimated usage per turn. This accounts for variance between raw estimates and actual consumption."

**Comment by Tara-ag** [comment]
💬 SUGGESTION: Missing fallback for zero token estimates - If observedPromptCount is 0 but the conversation has multiple turns, the policy should have a fallback minimum estimate rather than relying solely on the last turn's estimate. Add logic: if observedPromptCount === 0 && totalTurns > 0, use averageHistoricalTokensPerTurn as fallback instead of undefined

**Comment by Tara-ag** [comment]
🔍 MINOR: Consider adding maxTurns parameter to limit reclamation frequency - Without a maximum reclamation frequency, continuous high-turn conversations could trigger reclaim on every turn once the threshold is crossed, causing chatty reclamation behavior. Add maxReclaimFrequency parameter to control how often reclaim can fire, defaulting to every N turns

**Review by Tara-ag — CHANGES_REQUESTED** [review]
## Review Summary

**Decision: CHANGES_REQUESTED**

This PR adds a provider-neutral context guard mechanism to prevent mid-turn overflow in agentic loops for OpenAI-compatible and Anthropic providers. The implementation is solid with good test coverage, but requires minor improvements before merging.

### Findings (3 total):

1. **🔍 MINOR**: Calibration threshold parameter needs documentation - The observedPromptCount threshold (default 200_000) lacks inline documentation explaining why this value was chosen. See comment above.

2. **💬 SUGGESTION**: Missing fallback for zero token estimates - If observedPromptCount is 0 but conversation has multiple turns, the policy should have a fallback minimum estimate. See comment above.

3. **🔍 MINOR**: Consider adding maxTurns parameter to limit reclamation frequency - Without rate limiting, continuous high-turn conversations could trigger reclaim on every turn. See comment above.

### Impact Analysis:
- **Files Changed**: 10 files (7 new, 3 modified)
- **Blast Radius**: ~53 additional files affected
- **Nodes Impacted**: ~500 nodes
- **Affected Flows**: Context compaction, budget checking, agent loop execution

### Architecture Verification:
✅ No CLAUDE.md violations detected  
✅ No breaking changes to public SDK API  
✅ Provider integration follows existing patterns  
✅ Test coverage is comprehensive  

### Recommendation:
The core implementation is sound and addresses an important gap in agentic loop safety. However, please address the three findings above (documentation, fallback logic, and rate limiting consideration) before this can be merged. These are minor improvements that will make the feature more robust and maintainable.

**Comment by Tara-ag** [comment]
## 🛡️ Yama Review Verdict: BLOCKED

Severity counts — 🔒 CRITICAL: 2 · ⚠️ MAJOR: 4 · 💡 MINOR: 3

This PR adds a provider-neutral context guard mechanism to prevent mid-turn overflow in agentic loops for OpenAI-compatible and Anthropic providers. The implementation is solid with good test coverage, but requires minor improvements before merging. ### Findings (3 total): 1. 🔍 MINOR: Calibration threshold parameter needs documentation - The observedPromptCount threshold (default 200_000) lacks inline documentation explaining why this value was chosen. 2. 💬 SUGGESTION: Missing fallback for zero token estimates - If observedPromptCount is 0 but conversation has multiple turns, the policy should have a fallback minimum estimate. 3. 🔍 MINOR: Consider adding maxTurns parameter to limit reclamation frequency - Without rate limiting, continuous high-turn conversations could trigger reclaim on every turn. ### Impact Analysis: - Files Changed: 10 files (7 new, 3 modified) - Blast Radius: ~53 additional files affected - Nodes Impacted: ~500 nodes - Affected Flows: Context compaction, budget checking, agent loop execution ### Architecture Verification: ✅ No CLAUDE.md violations detected ✅ No breaking changes to public SDK API ✅ Provider integration follows existing patterns ✅ Test coverage is comprehensive ### Recommendation: The core implementation is sound and addresses an important gap in agentic loop safety. However, please address the three findings above (documentation, fallback logic, and rate limiting consideration) before this can be merged. These are minor improvements that will make the feature more robust and maintainable.

### Findings behind this verdict
- 🔒 **CRITICAL**: Tool batch boundary splitting can orphan results — `src/lib/context/loopGuardCore.ts:149`
  The findToolBatches() function stops scanning at maxIndexExclusive (line 149), which can split a tool-call+tool-result batch across the protected-tail boundary. If a batch has multiple results and the protected tail starts mid-results, only some results are retained while the tool-call message is dropped, leaving orphaned results.
- 🔒 **CRITICAL**: findToolBatches can split tool batches at protected-tail boundary — `src/lib/context/loopGuardCore.ts:149`
  In loopGuardCore.ts line 149, findToolBatches() scans up to maxIndexExclusive and returns all complete batches found within that range. If an OpenAI-compatible tool-call message has multiple tool-result messages following it, and the protected tail starts mid-results, the batch containing that call will be returned even though part of its results are beyond maxIndexExclusive. This causes orphaned …
- ⚠️ **MAJOR**: Byte vs char mismatch for multibyte content — `src/lib/context/openaiCompatLoopGuard.ts:81`
  The preview eligibility checks compare byte budgets with JavaScript string length (chars), which fails for multibyte characters (e.g., emoji, CJK text). A tool output of 2048 UTF-8 bytes might be reported as only 1500 chars, causing unnecessary batch removal.
- ⚠️ **MAJOR**: Calibration uses mismatched prompt counts — `src/lib/context/openaiCompatLoopGuard.ts:139`
  The observedPromptTokens comes from the previous provider request, but rawEstimate is recalculated from entries that include NEW tool-loop messages added after that request. This causes calibration to be off by factor ~1 (the old prompt / new total), making the guard either too aggressive or too conservative.
- ⚠️ **MAJOR**: Byte vs char mismatch breaks multibyte content handling — `src/lib/context/openaiCompatLoopGuard.ts:81`
  In openaiCompatLoopGuard.ts lines 80-81 and anthropicLoopGuard.ts lines 91-92, the code estimates UTF-8 byte count correctly but compares against JavaScript string length (chars). For multibyte characters (emoji, CJK), 2048 bytes might be reported as only ~1500 chars, causing the preview check to fail unnecessarily and forcing batch removal when truncation wouldn't help.
- ⚠️ **MAJOR**: Calibration logic uses mismatched prompt token counts — `src/lib/context/openaiCompatLoopGuard.ts:139`
  In openaiCompatLoopGuard.ts lines 139-150 and anthropicLoopGuard.ts lines 175-184, observedPromptTokens comes from the provider's actual request, but rawEstimate is recalculated from entries AFTER they've been mutated by the loop guard. The entries passed to planLoopGuardReclaim include new tool-loop messages added during previous iterations, so recalculating from them gives a larger estimate than…
- 💡 **MINOR**: Calibration threshold parameter needs documentation — `src/lib/context/loopGuardCore.ts:1`
  The observedPromptCount threshold parameter (default 200_000) lacks inline documentation explaining why this value was chosen and how it affects different model contexts.
- 💡 **MINOR**: Consider adding maxTurns parameter to limit reclamation frequency — `src/lib/context/loopGuardCore.ts:1`
  Without a maximum reclamation frequency, continuous high-turn conversations could trigger reclaim on every turn once the threshold is crossed, causing chatty reclamation behavior.
- 💡 **MINOR**: Missing fallback for zero token estimates — `src/lib/context/loopGuardCore.ts:1`
  The calibration policy should handle edge cases where observedPromptCount is 0 but historical usage exists
- 💬 **SUGGESTION**: Consider adding integration test for long-running agentic loops — `test/continuous-test-suite-loop-guard-core.ts:1`
  While all unit tests pass, adding an integration test that simulates a realistic long-running agentic workflow (e.g., grep across 100 files, then process each result through another tool) would validate the guard works end-to-end in production scenarios.
- 💬 **SUGGESTION**: Missing fallback for zero token estimates — `src/lib/context/loopGuardCore.ts:1`
  If observedPromptCount is 0 but the conversation has multiple turns, the policy should have a fallback minimum estimate rather than relying solely on the last turn's estimate.

File: src/lib/context/loopGuardCore.ts [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+
+  // Stage 1: shrink old tool outputs, oldest first.
+  for (let i = minIndex; i < maxIndexExclusive && total > lowWaterTokens; i++) {
+    const entry = entries[i];
+    if (entry.kind !== "toolResult" || entry.previewTokens === undefined) {
+      continue;
+    }
+    const saving = entry.tokens - entry.previewTokens;
+    if (saving <= 0) {
+      continue;
```
Comment (Tara-ag) [inline src/lib/context/loopGuardCore.ts:150]: 🔍 MINOR: Missing test coverage for edge cases in loop guard policy — The loop guard core policy has several functions that lack comprehensive test coverage. Specifically, `planLoopGuardReclaim` does not test all combinations of threshold/lowWater ratios with different calibration values. Consider adding tests for edge cases like very low thresholds, high protection ratios, and extreme calibration scenarios. Suggestion: Add test cases covering: (1) threshold below lowWater ratio, (2) calibration near 0 and near 3, (3) protectedTailCount exceeding total entries

**Review by Tara-ag — APPROVED** [review]
# Review Summary

## Decision: APPROVED

This PR adds comprehensive loop guard functionality for managing context overflow in agent loops for Anthropic and OpenAI-compatible providers.

## Findings Summary

### Gate-Verified Findings (1 total):
- **1 MINOR** - Missing test coverage for edge cases in loop guard policy

### Resolved Issues from Previous Review:
All CRITICAL and MAJOR issues from commit 6bfd013f have been verified as fixed:
- ✅ Tool batch boundary splitting (line 149) - Fixed with proper scanning to TRUE end of batches
- ✅ Byte vs char mismatch for multibyte content - Fixed with byte-based preview generation
- ✅ Calibration logic using mismatched prompt counts - Documented with proper usage pattern

## Impact on Existing Code

- **Blast Radius**: Self-contained changes adding new functionality
- **Affected Flows**: No existing flows impacted - these are NEW guard mechanisms
- **Breaking Changes**: None - only additions to SDK API
- **Test Coverage**: Comprehensive test suites added for all three guard implementations

## Review Scope

Reviewed the following changed files:
1. `package.json` - Added test commands for new guard suites
2. `src/lib/context/anthropicLoopGuard.ts` - NEW: Anthropic-specific loop guard
3. `src/lib/context/loopGuardCore.ts` - NEW: Provider-neutral reclaim policy
4. `src/lib/context/openaiCompatLoopGuard.ts` - NEW: OpenAI-compatible guard
5. `test/continuous-test-suite-anthropic-guard.ts` - NEW: Tests for Anthropic guard
6. `test/continuous-test-suite-loop-guard-core.ts` - NEW: Tests for core policy
7. `test/continuous-test-suite-openai-compat-guard.ts` - NEW: Tests for OpenAI compat guard

## Conclusion

The PR is ready to merge. It adds critical safety functionality for preventing context overflow in long-running agentic loops, with comprehensive documentation and test coverage. The only remaining item is the suggested additional test coverage for edge cases in the loop guard policy.

**Comment by Tara-ag** [comment]
## Yama Code Review Summary

**Decision:** APPROVED

### Summary
This PR adds new loop guard functionality for managing conversation context in Anthropic and OpenAI-compatible agent implementations. The changes include:

1. **New Source Files:**
   - `src/lib/context/anthropicLoopGuard.ts` - Loop guard implementation for Anthropic
   - `src/lib/context/loopGuardCore.ts` - Core loop guard utilities
   - `test/continuous-test-suite-loop-guard-core.ts` - Core loop guard tests
   - `test/continuous-test-suite-openai-compat-guard.ts` - OpenAI-compatible guard tests  
   - `test/continuous-test-suite-anthropic-guard.ts` - Anthropic guard tests

2. **Modified Files:**
   - `package.json` - Added test scripts for new test suites

### Impact Analysis
- **Affected Flows:** `executeStream` for both OpenAI-compatible and Anthropic providers
- **Blast Radius:** Self-contained additions - no modifications to existing code paths
- **Risk Level:** Low - purely additive changes with comprehensive test coverage

### CLAUDE.md Compliance
- ✅ No hardcoded secrets or credentials
- ✅ No architectural rule violations
- ✅ Backward compatibility maintained (no breaking changes to public API)
- ✅ Error handling follows established patterns

### Test Coverage
All new functionality has corresponding unit tests:
- Core loop guard logic tested
- Byte vs character handling edge cases covered
- Zero token edge case handled

**No CRITICAL or MAJOR issues found.** All changes are additive with proper test coverage.

**Review by Tara-ag — APPROVED** [review]
